lib/sensors/src/URM04Sensor.cpp

```diff
@@ -0,0 +1,197 @@
+    // counter
+    int RetryCounter{0};
+    // array to hold commands
+    uint8_t cmdrd[10];
+    // fill the array
+    for (int i{0}; i < 10; i++) {cmdrd[i] = 0;}
+    // start index & indices
+    int indices{0};
+    int start_index{0};
+    // flag to check if the indices in the buffer are valid
+    bool flag{true};
+    // variable to check if the whole buffer is valid
+    bool valid{false};
+    // header bit
+    uint8_t headerNo{0};
+
+    while (RetryCounter < MAX_TRIES && flag) {
+      // ensure serial is still running
+      if (nucleo_board.available()) {
+        // read from the serial port with BufferedSerial read and put  data into cmdrd
+        indices = nucleo_board.read(cmdrd, sizeof(cmdrd));
+        // did not read from serial port properly
+        if (indices == -1) {
+          flag = false;
+          // flush buffer
+          nucleo_board.sync();
+          // read failed
+          read_success = false;
+          break;  // exit loop
+        }
+        // confirm the header address is correct
+        if (cmdrd[start_index] == 0xAA) {
+          headerNo = 1;
+          valid    = true;
+        }
+        // confirm every bit is within the range of the buffer
+        if (valid && start_index == headerNo + 6) {
+          flag = false;
+          break;  // exit loop
+        }
+
+        start_index++;
+        RetryCounter = 0;
+
+      } else {
+        RetryCounter++;
+        ThisThread::sleep_for(std::chrono::microseconds(15));
+      }
+    }
+    // since all the data has proven to be valid send the data to be analyzed
+    if (valid) {
+      read_success = true;
+      analyzeUrmData(cmdrd);
+    }
+  }
+}
+// run the urm04
+void sensor::URM04Sensor::runUrm04() {
+  static uint64_t timer{0};
+  static int num{0};
+
+  if ((uint64_t)clock.read_ms() - timer > managerTimer) {
+    // write high to trip pin to turn on RS485 Transmitting mode
+    m_trigPin.write(1);
+
+    // state machine to read measurements from the sensor
+    switch (readingStep) {
+      case 0:
+        urmTrigger(urmID[num]);
+        managerTimer = 40;  // set interval for timed measurement after triggering measurements
+        break;
+      case 1:
+        urmReader(urmID[num]);
+        managerTimer = 0;  // set interval for measurements after reading distance command
+        break;
+      case 2:
+        m_trigPin.write(LOW);  // turn on reading mode for RS485
+        managerTimer = 10;
+        break;
+      default:
+        readingStep = 0;
+        break;
+    }
+
+    if (readingStep < 2)
+      readingStep++;
+    else
+      readingStep = 0;
+
+    timer = clock.read_ms();
+  }
+}
+// this function runs the sensor
+bool sensor::URM04Sensor::read() {
+  runUrm04();
+  decodeURM04();
+  return read_success;
+}
+// this function prints the data from the sensor
+void sensor::URM04Sensor::print_distance() {
+  for (int i{0}; i < urmAccount; i++) {
+    printf("Distance: %u\n", urmData[i]);
```

Comment:
this print statement doesnt make sesnse to me. u the output is:
Distance: some number
Distance: some other number.

which one is right? why are there 2?